test/widgets/message_list_test.dart

```diff
@@ -753,6 +753,96 @@ void main() {
           unreadMessageIds: [message.id]),
       ]);
 
+      group('MarkAsReadAnimation', () {
+        testWidgets('loading is changed correctly', (WidgetTester tester) async {
+          final narrow = TopicNarrow.ofMessage(message);
+          await setupMessageListPage(tester,
+            narrow: narrow, messages: [message], unreadMsgs: unreadMsgs);
+          check(isMarkAsReadButtonVisible(tester)).isTrue();
+
+          connection.prepare(json: UpdateMessageFlagsForNarrowResult(
+            processedCount: 11, updatedCount: 3,
+            firstProcessedId: null, lastProcessedId: null,
+            foundOldest: true, foundNewest: true).toJson());
+
+          final state = tester.state<MarkAsReadWidgetState>(find.byType(MarkAsReadWidget));
+          check(state.loading).isFalse();
+
+          await tester.tap(find.byType(MarkAsReadWidget));
+          await tester.pump();
+          check(state.loading).isTrue();
+
+          await tester.idle();
+          await tester.pump();
+          check(state.loading).isFalse();
+        });
```

Comment:
It feels like we should be able to test the user's experience of the loading state more directly. The `loading` field does power the user-visible signs that the button is loading, but it's also kind of an implementation detail. How about something like the following:

```diff
--- test/widgets/message_list_test.dart
+++ test/widgets/message_list_test.dart
@@ -1,4 +1,5 @@
 import 'dart:convert';
+import 'dart:ui';
 
 import 'package:checks/checks.dart';
 import 'package:collection/collection.dart';
@@ -888,6 +889,20 @@ void main() {
           unreadMessageIds: [message.id]),
       ]);
 
+      checkAppearsLoading(WidgetTester tester, bool expected) {
+        final semantics = tester.firstWidget<Semantics>(find.descendant(
+          of: find.byType(MarkAsReadWidget),
+          matching: find.byType(Semantics)));
+        check(semantics.properties.enabled).equals(!expected);
+
+        final opacity = tester.widget<AnimatedOpacity>(find.descendant(
+          of: find.byType(MarkAsReadWidget),
+          matching: find.byType(AnimatedOpacity)));
+        check(opacity.opacity).equals(expected ? 0.5 : 1.0);
+      }
+
       group('MarkAsReadAnimation', () {
         testWidgets('loading is changed correctly', (WidgetTester tester) async {
           final narrow = TopicNarrow.ofMessage(message);
@@ -895,21 +910,21 @@ void main() {
             narrow: narrow, messages: [message], unreadMsgs: unreadMsgs);
           check(isMarkAsReadButtonVisible(tester)).isTrue();
 
-          connection.prepare(json: UpdateMessageFlagsForNarrowResult(
-            processedCount: 11, updatedCount: 3,
-            firstProcessedId: null, lastProcessedId: null,
-            foundOldest: true, foundNewest: true).toJson());
+          connection.prepare(
+            delay: const Duration(milliseconds: 2000),
+            json: UpdateMessageFlagsForNarrowResult(
+              processedCount: 11, updatedCount: 3,
+              firstProcessedId: null, lastProcessedId: null,
+              foundOldest: true, foundNewest: true).toJson());
 
-          final state = tester.state<MarkAsReadWidgetState>(find.byType(MarkAsReadWidget));
-          check(state.loading).isFalse();
+          checkAppearsLoading(tester, false);
 
           await tester.tap(find.byType(MarkAsReadWidget));
           await tester.pump();
-          check(state.loading).isTrue();
+          checkAppearsLoading(tester, true);
 
-          await tester.idle();
-          await tester.pump();
-          check(state.loading).isFalse();
+          await tester.pump(const Duration(milliseconds: 2000));
+          checkAppearsLoading(tester, false);
         });
 
         testWidgets('loading is changed correctly if request fails', (WidgetTester tester) async {
```

Reply:
Thanks! I've created [a new NFC](https://github.com/zulip/zulip-flutter/pull/710/commits/55d3f861e48ee42b076e7c124ddcda22cf65d58a) for simulating delayed connections PTAL.

Reply:
oh.. I see there is an implementation for that merged already! rebasing..